src/Storages/RabbitMQ/RabbitMQBlockInputStream.cpp

```diff
@@ -0,0 +1,162 @@
+{
+    if (!claimed)
+        return;
+
+    storage.pushReadBuffer(buffer);
+}
+
+
+Block RabbitMQBlockInputStream::getHeader() const
+{
+    return metadata_snapshot->getSampleBlockForColumns(column_names, storage.getVirtuals(), storage.getStorageID());
+}
+
+
+void RabbitMQBlockInputStream::readPrefixImpl()
+{
+    auto timeout = std::chrono::milliseconds(context.getSettingsRef().rabbitmq_max_wait_ms.totalMilliseconds());
+
+    buffer = storage.popReadBuffer(timeout);
+    claimed = !!buffer;
+
+    if (!buffer || finished)
+        return;
+
+    buffer->checkSubscription();
+}
+
+
+Block RabbitMQBlockInputStream::readImpl()
+{
+    if (!buffer || finished)
+        return Block();
+
+    finished = true;
+
+    MutableColumns result_columns = non_virtual_header.cloneEmptyColumns();
+    MutableColumns virtual_columns = virtual_header.cloneEmptyColumns();
+
+    auto input_format = FormatFactory::instance().getInputFormat(
+            storage.getFormatName(), *buffer, non_virtual_header, context, 1);
+
+    InputPort port(input_format->getPort().getHeader(), input_format.get());
+    connect(input_format->getPort(), port);
+    port.setNeeded();
+
+    auto read_rabbitmq_message = [&]
+    {
+        size_t new_rows = 0;
+
+        while (true)
+        {
+            auto status = input_format->prepare();
+
+            switch (status)
+            {
+                case IProcessor::Status::Ready:
+                    input_format->work();
+                    break;
+
+                case IProcessor::Status::Finished:
+                    input_format->resetParser();
+                    return new_rows;
+
+                case IProcessor::Status::PortFull:
+                {
+                    auto chunk = port.pull();
+
+                    auto chunk_rows = chunk.getNumRows();
+                    new_rows += chunk_rows;
+
+                    auto columns = chunk.detachColumns();
+
+                    for (size_t i = 0, s = columns.size(); i < s; ++i)
+                    {
+                        result_columns[i]->insertRangeFrom(*columns[i], 0, columns[i]->size());
+                    }
+                    break;
+                }
+                case IProcessor::Status::NeedData:
+                case IProcessor::Status::Async:
+                case IProcessor::Status::Wait:
+                case IProcessor::Status::ExpandPipeline:
+                    throw Exception("Source processor returned status " + IProcessor::statusToName(status), ErrorCodes::LOGICAL_ERROR);
+            }
+        }
+    };
+
+    size_t total_rows = 0;
+
+    while (true)
+    {
+        if (buffer->eof())
+            break;
+
+        auto new_rows = read_rabbitmq_message();
+
+        auto exchange_name = buffer->getExchange();
+
+        for (size_t i = 0; i < new_rows; ++i)
+        {
+            virtual_columns[0]->insert(exchange_name);
+        }
+
+        total_rows = total_rows + new_rows;
+        buffer->allowNext();
```

Comment:
Suspicious. How buffer will work if we exit the loop in `->eof` branch? As I understand we reuse them.